WalletWasabi/Helpers/ByteArrayEqualityComparer.cs

```diff
@@ -0,0 +1,14 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Text;
+
+namespace WalletWasabi.Helpers
+{
+	public class ByteArrayEqualityComparer : IEqualityComparer<byte[]>
+	{
+		public bool Equals([AllowNull] byte[] x, [AllowNull] byte[] y) => ByteHelpers.CompareFastUnsafe(x, y);
```

Comment:
@kiminuo is adding `#nullable enable` and this is probably a good place to add it too instead of `[(Dis)AllowNullAttribute]`

Reply:
Yeah, that would be fantastic! I'll have a look what it would entail here - i.e. what warnings it would produce.

Reply:
(Forgot to mention the PR is here: #3870)

Reply:
https://github.com/kiminuo/WalletWasabi/commits/feature/2020-07-07-nullable-for-3869 if anything is useful, cherry-pick at your will.

Reply:
TBH I am not familiar with what I did here, I just followed the design of `StringNoWhiteSpaceEqualityComparer.cs`, which was created by @molnard: https://github.com/zkSNACKs/WalletWasabi/blob/master/WalletWasabi.Tests/Helpers/StringNoWhiteSpaceEqualityComparer.cs if I would've done it myself I would've do it differently, but I did not want to mix concepts, which brings me to @kiminuo's PR and your suggestion here.

I think @kiminuo's approach is a naive way to migrate the codebase to nullmagic. It's very noisy and mixes a bunch of concepts and eventually (even in this PR) new and new code will be written and the null magic won't be considered or added ever and we'll end up mixing concepts and making the codebase even harder to understand. I don't know how to migrate, but https://github.com/zkSNACKs/WalletWasabi/pull/3870 is not something that'll work IMO.

Reply:
@nopara73 I'm not familiar with "nullmagic" term but I'll try to respond somewhat verbosely: 

Nullable reference types (NRT) feature in C# compiler is based on the idea that most of the time you don't want your objects to be null and when you do, you should denote it in the declaration of the variable. This drastically reduces the need for null-checking (conditions whether variable is null) and gives code-reader confidence that she know whether given object can or cannot be null. 

So that's the motivation for the feature. You have your codebase and you can see in form of warnings that a variable is unassigned and that people who will touch that may get burned if it is not fixed.

Now, the question number one is whether you find this appealing. If not, then there is nothing to talk about and it's fine by me. I find `Nullable reference types` very useful in general but if others don't it's a lost battle so there is no need to start the battle :-)

If you think that NRT may be useful, then you need to consider how to start using them. There are two strategies:

1) Gradually add `#nullable enabled` pragmas to one file at the time.
2) Enable NRT at the project level. But there will be a lot of warnings.

You are right that the approach 1) is not forced upon people so it may be forgotten/abandoned etc. Interestingly, there are large projects like dotnet/roslyn (C# compiler) that precisely that (example https://github.com/dotnet/roslyn/pull/45640/files#diff-7fda8226c42aeb2350adc9ac491a6b1cR5). NRT is getting tracking quite nicely if you ask me. I have seen continued effort to add nullable annotations in dotnet/runtime too. 

Approach number 2 with enabling NRT at the project level is bit more drastic but each warning occurrence can be silenced by using `!` operator ([null-forgiving operator](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving)) so that it's immediately clear to a code-reader that something may not be entirely right.

So there are three choices:

1. Don't use NRT
2. Use NRT with gradual approach
3. Use NRT with project-level approach

PS: I don't do things naively.

Reply:
FYI Null Magic is the correct technical term :)

I am familiar with these approaches, but I am somewhat hoping there may be something between (2) and (3). Like ensuring that every new code must have comply with the approach. If it's not possible, then I would vote for (3), because (2) is pissing in the wind and (1) is admitting defeat.

Reply:
Personally, I like (3) the most. I just thought you guys would not support that.

Reply:
After release we talk about that and see if others are on board.

Reply:
Please bring it up in case we would forget it.

Reply:
I am with (3)